### ISSUE TITLE
Keep synced for connection losts

### ISSUE THREAD

**Issue by YerkoPalma** [yoshuawuyts/choo-persist#3]
This is like a feature request. Right now, this plugin allow a fast first load from indexDB, cool. But what happens when the connections is lost, and the app attemp to make an http request. It would be awesome to allow to the user to keep using the app, and save all changes to local db, and once the connection is availaible again, repost all the state to keep data synced.
Well, the thing is that I'm actually facing this problem right now, I'm writing a [pwa](https://developers.google.com/web/progressive-web-apps/) with choo [here](https://github.com/YerkoPalma/tic-tac-toe) and is working for now, but when I go offline, I have a fast first load, thanks to sw and idb, but can go any other place in the app because of xhr calls failling.
### Idea

I think this could be done using choo hooks. If there is a call for the `onError` hook due to xhr failing, save those changes only in the db, maybe an special collection only for offline diffs. Then on every state change, check if that special collection has data, in that case POST that data to sync the remote db, and clear that collection. 

Well that sounds simple but I know is not haha. I wanna hear your opinion about it. Does it make sense? Do you have any better idea?

**Comment by yoshuawuyts** [comment]
Interesting; this is def worth trying out I think.

Hmmm, I believe there's a built-in option in indexedDB to catch all requests while offline so they can be replayed, at the XHR level. Might be worth digging into.

I wonder if it wouldn't be easier to wrap `xhr` and prep it for replayability, as that might provide more control. Not sure; this is definitely worth experimenting with.

**Comment by rozap** [comment]
Seems like wrapping xhr at the application level and then hooking that to some sort of queue backed by choo-persist could be interesting. It'd be really cool if this wasn't xhr specific, lots of folks are doing rpc over a websocket. Maybe modeling this feature after an elm effect manager would allow it to be separated from the actual transport.

**Comment by YerkoPalma** [comment]
Exactly, in fact, even for my use case, keeping this feature xhr relative is an issue. The best approach would be to leave all the work for choo hooks and wrappers. It could also be done by [checking connectivity](https://davidwalsh.name/detecting-online) before getting any error. And using the same check, to sync data to the network when connection is availaible again. To make it more network agnostic, the user should provide a sync function so not only xhr is supported.
